### PR TITLE
Fix check of magic bytes

### PR DESCRIPTION
**Pull request by NiHoel**
Add unit test

### CODE REVIEW

File: src/qoi.cpp [hunk 1]
```diff
@@ -45,13 +45,15 @@ std::uint32_t read_4_be_bytes(std::vector<unsigned char> const& bytes, std::size
...
 bool is_valid(const std::vector<unsigned char>& bytes) noexcept {
-    if (bytes.size() < qoi::HEADER_SIZE) {
+    if (bytes.size() < qoi::HEADER_SIZE + qoi::END_MARKER_LENGTH) {
         return false;
     }
 
-    if (bytes[0] != QOI_MAGIC[0] and bytes[1] != QOI_MAGIC[1] and bytes[2] != QOI_MAGIC[3]) {
+    if (not std::equal(QOI_MAGIC.begin(), QOI_MAGIC.end(), bytes.begin()))
+        return false;
+
```
Comment (ShadowMitia) [inline src/qoi.cpp]: I prefer always having braces, even when they are optional.

```suggestion
    if (not std::equal(QOI_MAGIC.begin(), QOI_MAGIC.end(), bytes.begin())) {
        return false;
     }

```

File: src/qoi.cpp [hunk 2]
```diff
@@ -45,13 +45,15 @@ std::uint32_t read_4_be_bytes(std::vector<unsigned char> const& bytes, std::size
...
+	if (not std::equal(padding.rbegin(), padding.rend(), bytes.rbegin()))
         return false;
```
Comment (ShadowMitia) [inline src/qoi.cpp]: Same here

```suggestion
	if (not std::equal(padding.rbegin(), padding.rend(), bytes.rbegin())) {
        return false;
        }
```

File: tests/qoi_tests.cpp [hunk 3]
```diff
@@ -23,6 +23,12 @@ TEST_CASE("Header : get_header", "[header]") {
   REQUIRE(head.colorspace == 0);  
 }
 
+TEST_CASE("Header : is_valid", "[header]") {
+  std::vector<unsigned char> test_header{0x71, 0x6f, 0x69, 0x65, 0x00, 0x00, 0x03, 0x00, 0x0, 0x0, 0x2, 0x0, 0x3, 0x0, 0xfe, 0x74};
```
Comment (ShadowMitia) [inline tests/qoi_tests.cpp:27]: Test passes fine, but when I try to get the hex in an image and then run them through the C implementation it doesn't seem to be working...

**Review by ShadowMitia — CHANGES_REQUESTED** [review]
Thank you for your PR!

Overall it does seem ok, there's a couple of things to change, and you'll need to double check the test input, it might not be completely valid (but maybe there's a bug in the implementation?).

File: tests/qoi_tests.cpp (hunk 3 above)
Comment (NiHoel) [reply tests/qoi_tests.cpp:27]: The reference implementation should return NULL for this byte sequence. Does it decode it?

Comment (ShadowMitia) [reply tests/qoi_tests.cpp:27]: Nope, I just shouldn't do code reviews late at night :laughing:

**Review by ShadowMitia — APPROVED** [review]

**Comment by ShadowMitia** [comment]
Thank you for the changes!

**Pull request merged**